sql: add kv_catalog_* virtual tables

postamar opened the pull request:
This commit introduces the following virtual tables, which each mirror
more or less exactly a system table containing catalog information:
  - crdb_internal.kv_catalog_descriptor mirrors system.descriptor,
  - crdb_internal.kv_catalog_namespace mirrors system.namespace,
  - crdb_internal.kv_catalog_zones mirrors system.zones,
  - crdb_internal.kv_catalog_comments mirrors system.comments.

Each virtual table is constructed by overlaying the data in the system
table with the in-memory data in the transaction's descs.Collection.
This may include virtual object descriptors and comments as well.

This makes it possible to immediately get rid of the predefined_comments
virtual table and simply replace its usages with kv_catalog_comments.
Longer term, these new virtual tables are expected to become useful to
test declarative schema changer behavior, also they might prove
convenient for troubleshooting schema-related incidents because the
descriptors and zone configs are readily exposed as JSON values.

Informs #88571.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93336)
<!-- Reviewable:end -->

Commit:
schemadesc: endow synthetic schemas with protobufs

Synthetic schemas distinguish themselves by the fact that they are not
persisted in system.descriptor and therefore do not have protobuf
descriptors.

That being said, we should be able to generate said protobuf descriptors
on demand for the sake of observability.

Release note: None

Commit:
sql: add crdb_internal.kv_catalog_* virtual tables

This commit introduces the following virtual tables, which each mirror
more or less exactly a system table containing catalog information:

crdb_internal.kv_catalog_descriptor mirrors system.descriptor,
crdb_internal.kv_catalog_namespace mirrors system.namespace,
crdb_internal.kv_catalog_zones mirrors system.zones,
crdb_internal.kv_catalog_comments mirrors system.comments.
Each virtual table is constructed by overlaying the data in the system
table with the in-memory data in the transaction's descs.Collection.
This may include virtual object descriptors and comments as well.

This makes it possible to immediately get rid of the predefined_comments
virtual table and simply replace its usages with kv_catalog_comments.
Longer term, these new virtual tables are expected to become useful to
test declarative schema changer behavior, also they might prove
convenient for troubleshooting schema-related incidents because the
descriptors and zone configs are readily exposed as JSON values.

Informs #88571.

Release note: None

ajwerner commented on pkg/sql/crdb_internal.go:
```diff
@@ -4895,51 +4900,154 @@ CREATE TABLE crdb_internal.kv_store_status (
+		// Delegate privilege check to system table.
+		{
+			sysTable := all.LookupDescriptor(systemschema.DescriptorTable.GetID())
+			if p.CheckAnyPrivilege(ctx, sysTable) != nil {
```
Should this be CheckAnyPrivilege or should it be something else? Namely select

postamar replied:
Oh, good point. I shouldn't have blindly copied that.

ajwerner reviewed (approved):
LGTM

postamar commented on pkg/sql/catalog/schemadesc/public_schema_desc.go:
```diff
@@ -42,25 +42,21 @@ type public struct {
-	return privilege.Schema
-}
+func (p public) GetID() descpb.ID                     { return keys.PublicSchemaID }
+func (p public) GetParentID() descpb.ID               { return descpb.InvalidID }
```
No, there's a mildly annoying edge case involving restores, something about the temporary system database also not having a public schema descriptor.

postamar commented:
Thanks for the review!

bors r+
